### PR TITLE
Fix some layering violations and add some type annotations in the interpreter

### PR DESCRIPTION
**Pull request by dcbaker**
This was peeled out of my rust module series, as it's not strictly related, and should be easy enough to review and land now.

### CODE REVIEW

File: mesonbuild/build.py [hunk 1]
```diff
@@ -140,9 +187,9 @@ def __init__(self, environment: environment.Environment):
         self.projects_link_args = PerMachine({}, {})  # type: PerMachine[T.Dict[str, T.List[str]]]
         self.tests = []                               # type: T.List['Test']
         self.benchmarks = []                          # type: T.List['Test']
-        self.headers = []
-        self.man = []
-        self.data = []
+        self.headers: T.List[Headers] = []
+        self.man: T.List[Man] = []
+        self.data = []                                # type: T.List[Data]
```
Comment (mensinda) [inline mesonbuild/build.py:192]: This should also be the new syntax now...

**Review by mensinda — COMMENTED** [review]
LGTM, besides the one minor issue below.

File: mesonbuild/build.py (hunk 1 above)
Comment (dcbaker) [reply mesonbuild/build.py]: Ah, good catch. I've fixed that, and updated the Build initializer to use only the new syntax for consistency.

**Review by mensinda — APPROVED** [review]
LGTM, assuming the CI is green

**Pull request merged**